### PR TITLE
Add codegen test for array comparision opt

### LINKED ISSUE #62531: Missed optimization on array comparison
The godbolt link: https://godbolt.org/z/dc9o3x

I think this snippet should just return `true`:
```rust
pub fn compare() -> bool {
    let bytes = 12.5f32.to_ne_bytes();
    bytes == if cfg!(target_endian = "big") {
        [0x41, 0x48, 0x00, 0x00]
    } else {
        [0x00, 0x00, 0x48, 0x41]
    }
}
```

The generated asm with `opt-level=3`:
```asm
example::compare:
        push    rax
        mov     al, 1
        pop     rcx
        ret
```

### PR DESCRIPTION
**Pull request by tesuji**
Closes #62531

r? @jonas-schievink

### CODE REVIEW

**Comment by tesuji** [comment]
note for self: use min-llvm

File: src/test/assembly/62531-array-cmp.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// Test to make sure that asm for array comparison on x86 optimized.
```
Comment (LingMan) [inline src/test/assembly/62531-array-cmp.rs:1]:
```suggestion
// Ensure the asm for array comparisons is properly optimized on x86.
```

**Comment by Dylan-DPC** [comment]
@jonas-schievink this is ready for review

**Comment by jonas-schievink** [comment]
I don't really know how this works.

r? @nagisa

File: src/test/assembly/62531-array-cmp.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+// min-llvm-version: 11
+// assembly-output: emit-asm
+// compile-flags: -C opt-level=2
+// only-x86_64
+
+#![crate_type = "lib"]
+
+// CHECK-LABEL: compare
+// CHECK: movb $1, %al
```
Comment (nagisa) [inline src/test/assembly/62531-array-cmp.rs:11]: Probably should `CHECK-NEXT` here. There are a number of plausible ways to generate code for this that would result in these two instructions somewhere, but not necessarily in _just_ these two instructions.

File: src/test/assembly/62531-array-cmp.rs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+// CHECK-NEXT: retq
+#[no_mangle]
+pub fn compare() -> bool {
+    let bytes = 12.5f32.to_ne_bytes();
+    bytes == if cfg!(target_endian = "big") {
```
Comment (nagisa) [inline src/test/assembly/62531-array-cmp.rs:16]: This is a `x86_64`-only test, so I don't think this branch is relevant?

Comment (nagisa) [reply src/test/assembly/62531-array-cmp.rs:16]: Ah, never mind, this is checking if we end up optimizing a unconditional branch, I guess.

**Comment by nagisa** [comment]
This is r=me once the comment above is addressed.

**Comment by JohnCSimon** [comment]
Ping from triage: @tesuji  can you post your status on this MR and address nagisa's feedback? Thank you.

**Comment by nagisa** [comment]
Alas the issue seems to have resurfaced. I rebased the test case and it failed. Closing this to get it out of my queue.

**Pull request closed without merging**